lib/mix/lib/mix/text_diff.ex

```diff
@@ -0,0 +1,334 @@
+      iex> formatted = code |> Code.format_string!() |> IO.iodata_to_binary()
+      iex> code
+      ...> |> Mix.TextDiff.format(formatted, color: false)
+      ...> |> IO.iodata_to_binary()
+      """
+       1  1   |defmodule Foo do
+       2    - |  @moduledoc   false
+          2 + |  @moduledoc false
+       3  3   |
+       4    - |  def foo, do:  :foo
+          4 + |  def foo, do: :foo
+       5  5   |
+       6  6   |  def three_times(x) do
+       7    - |    {x,
+       8    - |     x,x}
+          7 + |    {x, x, x}
+       9  8   |  end
+      10  9   |
+           ...|
+      """
+  '''
+
+  @newline "\n"
+  @blank " "
+
+  @separator "|"
+  @line_num_pad @blank
+
+  @gutter [
+    del: " - ",
+    eq: "   ",
+    ins: " + ",
+    skip: "..."
+  ]
+
+  @colors [
+    del: [text: :red, space: :red_background],
+    ins: [text: :green, space: :green_background]
+  ]
+
+  @default_opts [
+    after: 2,
+    before: 2,
+    color: true,
+    line: 1
+  ]
+
+  @doc ~S'''
+  Formats the diff between two strings.
+
+  The returned `iodata` shows the lines with changes and 2 lines before and
+  after the changed positions. The string contains also a gutter with line
+  number and a `-` or `+` for removed and added lines. Multiple lines without
+  changes are marke with `...` in the gutter.
+
+  ## Options
+
+    * `after` - the count of lines printed after each change. Defaults to `2`.
+    * `before` - the count of lines printed befor each change. Defaults to `2`.
+    * `color` - enables color in the output. Defautls to `truel.`
+    * `line` - the line number of the first line. Defaults to `1`.
+
+  ## Examples
+
+      iex> code = """
+      ...> defmodule Bar do
+      ...>   @moduledoc false
+      ...>
+      ...>   bar(x, y) do
+      ...>     z = x + y
+      ...>     {x,y  , z}
+      ...>   end
+      ...>
+      ...>   bar(x, y, z) do
+      ...>     {x, y, z}
+      ...>   end
+      ...> end\
+      ...> """
+      iex> formatted = code |> Code.format_string!() |> IO.iodata_to_binary()
+      iex> code
+      ...> |> Mix.TextDiff.format(formatted, color: false)
+      ...> |> IO.iodata_to_binary()
+      """
+           ...|
+       4  4   |  bar(x, y) do
+       5  5   |    z = x + y
+       6    - |    {x,y  , z}
+          6 + |    {x, y, z}
+       7  7   |  end
+       8  8   |
+           ...|
+      """
+      iex> code
+      ...> |> Mix.TextDiff.format(formatted, color: false, after: 1, before: 1)
+      ...> |> IO.iodata_to_binary()
+      """
+           ...|
+       5  5   |    z = x + y
+       6    - |    {x,y  , z}
+          6 + |    {x, y, z}
+       7  7   |  end
+           ...|
+      """
+  '''
+  @spec format(String.t(), String.t(), keyword()) :: iodata()
+  def format(code, code, opts \\ default_opts())
+
+  def format(code, code, _opts), do: []
+
+  def format(old, new, opts) do
+    opts = Keyword.merge(@default_opts, opts)
+
+    old = String.split(old, "\n")
+    new = String.split(new, "\n")
```

Comment:
```suggestion
    old = String.split(old, ["\r\n", "\n"])
    new = String.split(new, ["\r\n", "\n"])
```

Reply:
But then what happens when two files are equal except for `\r\n` -> `\n`? So maybe it is better to keep it by your current split?

Reply:
Yes, splitting with `\n` should be better. The last commit makes also changed `\r` visible.
```
iex(1)> Mix.TextDiff.format("a\r\nb\r\n", "a\nb\nc\n") |> IO.iodata_to_binary() |> IO.puts()
1   - |a↵
2   - |b↵
  1 + |a
  2 + |b
  3 + |c

:ok
iex(2)> Mix.TextDiff.format("a\r\n", "a\n") |> IO.iodata_to_binary() |> IO.puts()
1   - |a↵
  1 + |a

:ok
iex(3)> Mix.TextDiff.format("a\n", "a\r\n") |> IO.iodata_to_binary() |> IO.puts()
1   - |a
  1 + |a↵

:ok
```